jfr/src/main/java/io/netty/handler/jfr/JfrEventHandler.java

```diff
@@ -0,0 +1,375 @@
+            event.remoteAddress = String.valueOf(remoteAddress);
+            event.localAddress = String.valueOf(localAddress);
+            event.commit();
+        }
+    }
+
+    @Override
+    public void disconnect(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        DisconnectEvent event = new DisconnectEvent();
+        event.begin();
+        ctx.disconnect(promise);
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        CloseEvent event = new CloseEvent();
+        event.begin();
+        ctx.close(promise);
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+    @Override
+    public void deregister(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        DeregisterEvent event = new DeregisterEvent();
+        event.begin();
+        ctx.deregister(promise);
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+// NOTE: Occurs so often, currently remove for perfomance reasons
+//    @Override
+//    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+//        ChannelReadCompleteEvent event = new ChannelReadCompleteEvent();
+//        event.begin();
+//        ctx.fireChannelReadComplete();
+//        event.end();
+//        if (event.shouldCommit()) {
+//            event.setChannel(ctx.channel());
+//            event.commit();
+//        }
+//    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        ChannelReadEvent event = new ChannelReadEvent();
+        event.begin();
+        ctx.fireChannelRead(msg);
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.msg = String.valueOf(msg);
+            event.commit();
+        }
+    }
+
+// NOTE: Occurs so often, currently remove for perfomance reasons
+//    @Override
+//    public void read(ChannelHandlerContext ctx) throws Exception {
+//        ReadEvent event = new ReadEvent();
+//        event.begin();
+//        ctx.read();
+//        event.end();
+//        if (event.shouldCommit()) {
+//            event.setChannel(ctx.channel());
+//            event.commit();
+//        }
+//    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        WriteEvent event = new WriteEvent();
+        event.begin();
+        ctx.write(msg, promise);
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.msg = String.valueOf(msg);
```

Comment:
This is not really safe imho... The `msg` may have been released at this point already. You will need to convert it to a string before calling `ctx.write(...)`.